lib/galaxy/datatypes/annotation.py

```diff
@@ -54,15 +54,15 @@ class Augustus(CompressedArchive):
     edam_data = "data_0950"
     compressed = True
 
-    def set_peek(self, dataset: "DatasetInstance", **kwd) -> None:
+    def set_peek(self, dataset: DatasetProtocol23, **kwd) -> None:
```

Comment:
@jdavcs Thanks for all this work.

I would have started with a wide Protocol for all of this and then narrowed it down if a distinction made sense for display_peek versus generate_primary_file, etc... But you've gone... too far here I think. I hope that is fair.

These protocols don't have useful names and are describing what the method is doing - not the abstraction that a consumer of this module would need to provide. You're doing the work the just in time compiler - not providing good source documentation? To be more useful and more straightforward - I think you should condense the Protocol into the widest things that is useful for all the set_peek/display_peek functions call that Peekable(), widest thing for all the generate_primary_file() and call that ContainsExtraFiles(), etc... These are the types that new datatype implementers should be writing against and that consumers should provide an implementation of - and I think that this is the utility of that type system?

If there are obscure features only used in one or two methods - you can document that in Protocol.

Reply:
Thank you for the detailed feedback, @jmchilton - it's most helpful! I'll rework this. 
This, in particular -  

> These are the types that new datatype implementers should be writing against and that consumers should provide an implementation of - and I think that this is the utility of that type system?

makes perfect sense.

Reply:
@jmchilton So, this is my attempt at collapsing Protocols into logical sets of attributes. The last intermediate step was [here](https://github.com/galaxyproject/galaxy/blob/399da08bf9b56024a63e3c7c4b804de12688460a/lib/galaxy/datatypes/_protocols.py): that's the `_protocols` module in pre-final-collapse state. I first grouped protocol attributes based on what the methods required (start with a callable that's called only from outside the datatypes package and has a dataset parameter, then recursively update signatures of the callables that are called from that initial method/function, combining required attributes into one protocol; then move on to the next callable). That resulted in protocols that, of course, described what those methods were doing. I tried collapsing that into semantically-meaningful groups, but couldn't come up with anything useful. So I ended up with this: (1) a base dataset protocol; (2) base + hid, for HDAs, (3); a protocol for generating a primary file, which can be passed a model instance, or a Bunch, so it's much narrower; and (4) several individual mini-protocols (`HasFoo`) which are part of the base protocol, but are referenced in the datatypes modules where a method might only need an object that has a `Foo`.

What do you think?

Reply:
I don't have a complete understanding of this project yet, but intuitively I would prefer treating datatypes more like tools i.e. with a clearly specified interface (possibly even using `yml` files?), instead of attempting to narrowly type module internals.